### PR TITLE
feat(source): support json schema

### PR DESCRIPTION
**Pull request by wugouzi**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Support json schema as schema registry. Syntax is 
``` 
create table t with (
  connector = 'kafka',
  topic = 'json_schema',
  properties.bootstrap.server = 'message_queue:29092',
  scan.startup.mode = 'earliest'
) FORMAT PLAIN ENCODE JSON (schema.registry = 'url')
```
Basic idea is to use [jsonschema-transpiler](https://github.com/mozilla/jsonschema-transpiler) to generate avro schema from json schema.

Redpanda does not support json schema yet (see https://github.com/redpanda-data/redpanda/issues/1878). So kafka is used here as workaround. 

[References outside the schema](https://json-schema.org/learn/getting-started-step-by-step#references-outside-the-schema) is not supported yet.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by StrikeW** [comment]
Is schema registry is required for JSON format after this pr? IIRC, we have a conclusion before that schema registry is required for Avro format. cc @neverchanje

**Comment by wugouzi** [comment]
> Is schema registry is required for JSON format after this pr? IIRC, we have a conclusion before that schema registry is required for Avro format. cc @neverchanje

It's an option. We still support JSON format in old way.

**Comment by tabVersion** [comment]
> Is schema registry is required for JSON format after this pr?

It is not required. We implement it at a poc user's request.

File: src/connector/src/parser/json_parser.rs [hunk 1]
```diff
@@ -164,6 +226,16 @@ mod tests {
...
 
+    fn get_json_config() -> SpecificParserConfig {
+        SpecificParserConfig {
+            key_encoding_config: None,
+            encoding_config: EncodingProperties::Json(JsonProperties {
+                use_schema_registry: false,
+            }),
+            protocol_config: ProtocolProperties::Plain,
+        }
+    }
```
Comment (tabVersion) [inline src/connector/src/parser/json_parser.rs:237]: may try

```
impl SpecificParserConfig {
  const JSON = SpecificParserConfig { ... }
}
```

it is not a good idea to leave a func

File: src/connector/src/parser/maxwell/simd_json_parser.rs [hunk 2]
```diff
@@ -35,7 +35,9 @@ mod tests {
 
         let props = SpecificParserConfig {
             key_encoding_config: None,
-            encoding_config: EncodingProperties::Json(JsonProperties {}),
+            encoding_config: EncodingProperties::Json(JsonProperties {
+                use_schema_registry: false,
+            }),
```
Comment (tabVersion) [inline src/connector/src/parser/maxwell/simd_json_parser.rs:40]: why set to false?

File: src/connector/src/parser/mod.rs [hunk 3]
```diff
@@ -884,15 +886,19 @@ impl SpecificParserConfig {
...
                 | SourceFormat::DebeziumMongo
                 | SourceFormat::Maxwell
                 | SourceFormat::Canal
-                | SourceFormat::Plain
                 | SourceFormat::Upsert,
                 SourceEncode::Json,
-            ) => EncodingProperties::Json(JsonProperties {}),
+            ) => EncodingProperties::Json(JsonProperties {
+                use_schema_registry: false,
+            }),
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:901]: why cannot they have schema registry?

File: src/frontend/src/handler/create_source.rs [hunk 4]
```diff
@@ -225,6 +225,21 @@ fn consume_string_from_options(
     ))))
 }
 
+fn get_json_schema_location(
+    row_options: &mut BTreeMap<String, String>,
```
Comment (tabVersion) [inline src/frontend/src/handler/create_source.rs:229]: can we reuse the avro one?

File: src/frontend/planner_test/tests/testdata/output/create_source.yaml [hunk 5]
```diff
@@ -6,7 +6,7 @@
 - id: create_source_without_connector
   sql: |
     create source s() with(a=1) FORMAT PLAIN ENCODE JSON;
-  planner_error: 'Invalid input syntax: schema definition is required for ENCODE JSON'
+  planner_error: 'Protocol error: missing field ''connector'''
```
Comment (tabVersion) [inline src/frontend/planner_test/tests/testdata/output/create_source.yaml:9]: @st1page please double check the change.

File: src/frontend/src/handler/create_source.rs [hunk 6]
```diff
@@ -331,15 +346,52 @@ pub(crate) async fn try_bind_columns_from_source(
...
-                ..Default::default()
-            },
-        ),
+        (Format::Plain, Encode::Json) => {
+            let schema_config = get_json_schema_location(&mut options)?;
+            if schema_config.is_some() && sql_defined_schema {
+                return Err(RwError::from(ProtocolError(
+                    "User-defined schema is not allowed with schema registry.".to_string(),
+                )));
+            }
```
Comment (tabVersion) [inline src/frontend/src/handler/create_source.rs:355]: what about watermark?

**Comment by tabVersion** [comment]
> [References outside the schema](https://json-schema.org/learn/getting-started-step-by-step#references-outside-the-schema) is not supported yet.

I don't see how you handle this situation.

File: src/connector/src/parser/json_parser.rs [hunk 7]
```diff
@@ -113,6 +140,40 @@ impl JsonParser {
...
+            "https" | "http" => read_schema_from_http(&url).await,
+            scheme => Err(RwError::from(ProtocolError(format!(
+                "path scheme {} is not supported",
+                scheme
+            )))),
+        }?
+    };
+    let json_schema = serde_json::from_str(&schema_content)?;
+    let context = Context::default();
+    let avro_schema = convert_avro(&json_schema, context).to_string();
```
[the remark this answers is not in the document]
Comment (wugouzi) [reply src/connector/src/parser/json_parser.rs:171]: Yes the json schema is converted to avro schema and then we infer the columns from it.

**Comment by wugouzi** [comment]
> > [References outside the schema](https://json-schema.org/learn/getting-started-step-by-step#references-outside-the-schema) is not supported yet.
> 
> I don't see how you handle this situation.

[jsonschema-transpiler](https://github.com/mozilla/jsonschema-transpiler) does not support this based on my testing.

File: src/connector/src/parser/maxwell/simd_json_parser.rs (hunk 2 above)
Comment (wugouzi) [reply src/connector/src/parser/maxwell/simd_json_parser.rs:40]: this is for test

File: src/frontend/src/handler/create_source.rs (hunk 4 above)
Comment (wugouzi) [reply src/frontend/src/handler/create_source.rs:229]: nope since `(None,None)` is valid for json schema.

**Comment by wugouzi** [comment]
Performance
```
json_parser             time:   [1.7250 s 1.7257 s 1.7266 s]
                        change: [-0.3861% -0.3042% -0.2271%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

**Review by tabVersion — APPROVED** [review]
basically LGTM, waiting @st1page for syntax check.

**Review by st1page — APPROVED** [review]

**Comment by BugenZhao** [comment]
This PR introduces too many outdated and unnecessary dependencies to our workspace, including `env_logger` and `clap 2`. We should consider forking the dependency of the executable `jsonschema-transpiler` and making it a modern library.

**Pull request merged**